add an indexing# namespace for indexing-related values in fedora

acoburn opened the pull request:
https://jira.duraspace.org/browse/FCREPO-1285

This is a straw-man implementation of a http://fedora.info/definitions/v4/config# namespace

Comments welcome!

escowles commented on config.rdf:
```diff
@@ -0,0 +1,27 @@
+        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
+    </owl:Ontology>
+
+    <owl:ObjectProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransform">
```
Shouldn't this be a DatatypeProperty?  The range is literal, and in the current implementation it is a path segment treated as a string.

acoburn replied:
Yes, I believe it should be a DatatypeProperty. Also, the range should be `rdf:resource="&xsd;string"`

escowles replied:
:+1:  sounds right to me.

acoburn replied:
Also, this property should be `hasIndexingTransformation` as per the documentation: https://wiki.duraspace.org/display/FEDORA40/External+Search

acoburn commented on config.rdf:
```diff
@@ -0,0 +1,27 @@
+
+    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
+        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
+        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
```
should the domain for the hasIndexingTransformation property be repository:Container or repository:Resource? Practically speaking, it seems that repository:Container is correct, but should it be less restrictive?

ajs6f replied:
The correct domain is "indexing:Indexable", or "config:Indexable", which is what it seems to have been called here.

escowles replied:
Wait, the implication here is that the only reason to transform something is to index it.  Is that truly the only reason?  Or are we open to transforming Containers that aren't Indexable for other purposes?

ajs6f replied:
No, the implication is that the only reason that something possesses an "indexing transformation" is because it is indexable. We have no language for general "tranformability" because, under the current rules of the transform API, any resource with an RDF representation available at its URI is transformable, at least as I understand the matter. We may wish to introduce such language, but that's a separate concern from "indexability". In fact, we could probably profit from revisiting the transform API and trying to generalize it.

Incidentally, this kind of confusion is why I voted against compressing "auxiliary" predicates and types into a grab-bag "config" namespace.

awoods replied:
Did I miss your vote against a config# namespace here: https://github.com/fcrepo4/ontology/issues/12 ?

ajs6f replied:
Yes, I guess you did.


On Thursday, January 15, 2015, Andrew Woods <notifications@github.com>
wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23055684>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> Did I miss your vote against a config# namespace here: #12
> <https://github.com/fcrepo4/ontology/issues/12> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23055684>.
>

awoods replied:
@ajs6f, Did you have a counter-proposal?

ajs6f replied:
An indexing namespace specific to the purpose...

On Friday, January 16, 2015, Andrew Woods <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23084331>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> @ajs6f <https://github.com/ajs6f>, Did you have a counter-proposal?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23084331>.
>

awoods replied:
We need to get resolution on this. The question on the table is:
`Do we have one namespace that defines the predicates that users can update on resources, or do we have multiple such namespaces?`
Does the potential clarity provided by multiple namespaces outweigh the maintenance issues and the potential user-confusion that may be created in having to look for the right namespace?

ajs6f replied:
Well, my opinion is obviously for modularity.

On Friday, January 16, 2015, Andrew Woods <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23088144>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> We need to get resolution on this. The question on the table is:
> Do we have one namespace that defines the predicates that users can update
> on resources, or do we have multiple such namespaces?
> Does the potential clarity provided by multiple namespaces outweigh the
> maintenance issues and the potential user-confusion that may be created in
> having to look for the right namespace?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23088144>.
>

escowles replied:
I'm in favor of a single namespace for all new fedora-defined user-editable properties.  I think having multiple namespaces just makes things more cumbersome for the user, and we can name and describe the new properties well enough to make it clear what they are for.

awoods replied:
@acoburn? @escowles? opinions?

ajs6f replied:
When you write "Fedora-defined", do you mean Fedora the repository or
Fedora the software project? Because the indexing gear is defined by the
latter, not the former.

On Friday, January 16, 2015, Esmé Cowles <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23088452>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> I'm in favor of a single namespace for all new fedora-defined
> user-editable properties. I think having multiple namespaces just makes
> things more cumbersome for the user, and we can name and describe the new
> properties well enough to make it clear what they are for.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23088452>.
>

acoburn replied:
My main interest is in having a published ontology.
I don't have a strong opinion on whether this goes into an indexing# or a config# namespace. It seems that a single namespace makes management a lot easier, but it also introduces a certain degree of messiness in the ontology.

escowles replied:
@ajs6f I mean Fedora-the-project.  I know that namespaces are cheap, and I take the point that there is a certain tidiness of having each concern have its own ontology.  But as a counter-point, I think developers are better served by having a single place to look for properties they can set to trigger repository/indexer/etc. behavior.

ajs6f replied:
We can have both, with proper ontology design.

On Friday, January 16, 2015, Esmé Cowles <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23094609>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> @ajs6f <https://github.com/ajs6f> I mean Fedora-the-project. I know that
> namespaces are cheap, and I take the point that there is a certain tidiness
> of having each concern have its own ontology. But as a counter-point, I
> think developers are better served by having a single place to look for
> properties they can set to trigger repository/indexer/etc. behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23094609>.
>

awoods replied:
@ajs6f, could you please elaborate?

ajs6f replied:
A separate indexing ontology could and should import the core ontology.
Only one dependency would appear to the integrator.

On Friday, January 16, 2015, Andrew Woods <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23098473>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> @ajs6f <https://github.com/ajs6f>, could you please elaborate?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23098473>.
>

awoods replied:
@ajs6f:
@acoburn has already done the work:
https://github.com/fcrepo4/ontology/pull/13
https://github.com/fcrepo4/fcrepo-camel/pull/38
https://github.com/fcrepo4/fcrepo-message-consumer/pull/67

Is your suggestion a matter of preference, or do you see there being a more fundamental, and eventually painful, downside to having a single `config#` namespace. If the former, I would suggest we let implementation trump personal preference.

ajs6f replied:
1. That is exactly the opposite of what I am suggesting.

2. This is not a matter of preference of any kind.

Shall we move this to the next committers' meeting?

On Friday, January 16, 2015, Andrew Woods <notifications@github.com> wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23113037>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> @ajs6f <https://github.com/ajs6f>:
> @acoburn <https://github.com/acoburn> has already done the work:
> #13 <https://github.com/fcrepo4/ontology/pull/13>
> fcrepo4/fcrepo-camel#38 <https://github.com/fcrepo4/fcrepo-camel/pull/38>
> fcrepo4/fcrepo-message-consumer#67
> <https://github.com/fcrepo4/fcrepo-message-consumer/pull/67>
>
> Is your suggestion a matter of preference, or do you see there being a
> more fundamental, and eventually painful, downside to having a single
> config# namespace. If the former, I would suggest we let implementation
> trump personal preference.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23113037>.
>

awoods replied:
@ajs6f, There are two suggestions on the table, single namespace or multiple namespaces. The single namespace approach has pull-requests associated with it.
1. I am aware that the pull-requests do not represent your suggested approach.
2. If this is not a matter of preference, my question remains: Do you have any additional arguments for a multiple namespace approach beyond what you have already written in these comments?

It is always a good idea to bring more of the community into discussions. I expect that arguments presented at the tech meeting could also be expressed here.

acoburn replied:
Well, please don't decide on the single namespace approach just because there is an existing pull request. The associated code changes in fcrepo-camel and fcrepo-message-consumer are truly minimal.

I would only add that both Stefano and I have use cases for the *:Indexable class and *:hasIndexingTransform property, and the PR was primarily an attempt to begin a discussion so that we can move things along.

@ajs6f I am not particularly familiar with the use of the [owl:imports](http://www.w3.org/TR/owl-ref/#imports-def) statement. If A imports B, does that mean that all of the classes and properties of B are now available under the namespace of A?

ajs6f replied:
There seems to be some confusion around the idea that a single-namespace
approach will create less work for clients. It won't. In any event, no one
will have to check or use more than one namespace.

More importantly, the single-namespace approach is wrong. It misunderstands
the upcoming spec'ing and partitioning of the API, and the concomitant
approach to the use of ontology. The predicates involved in this example
(the indexing stuff) have literally _nothing_ to do with the operation of
any repository. _Nothing_. They concern _only_ the operation of the
indexing apparatus. So they are naturally associated _to that machinery_.
But there's more to be said.

More broadly, the only ontologies that should be showing up in this
repository are those that would be reused by a completely fresh
implementation of the Fedora repository architecture. (E.g. at UCSD or
Oxford.) It's great that we want to use strong declarative languages like
RDFS and OWL to explain configuration for our implementation. I'm all in
favor of that. But those ontologies don't belong here. They belong in our
implementation, under a namespace specific to our implementation.
(Obviously, they should be _published_ elsewhere, but I'm talking about
where they are maintained.) The indexing predicates should be maintained,
similarly, _with the indexing machinery code_. (Unless we expect to
standardize an indexing SPI to complement the repository APIs.)

Putting a "config" namespace here confuses the Fedora repository
architecture with our implementation thereof, and putting the indexing
predicates into it confuses the repository with the indexer. We can't
expect our community to develop any understanding if we keep confusing them
like that.



On Fri, Jan 16, 2015 at 7:54 PM, Aaron Coburn <notifications@github.com>
wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23119962>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> Well, please don't decide on the single namespace approach just because
> there is an existing pull request. The associated code changes in
> fcrepo-camel and fcrepo-message-consumer are truly minimal.
>
> I would only add that both Stefano and I have use cases for the
> *:Indexable class and *:hasIndexingTransform property, and the PR was
> primarily an attempt to begin a discussion so that we can move things along.
>
> @ajs6f <https://github.com/ajs6f> I am not particularly familiar with the
> use of the owl:imports <http://www.w3.org/TR/owl-ref/#imports-def>
> statement. If A imports B, does that mean that all of the classes and
> properties of B are now available under the namespace of A?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23119962>.
>

escowles replied:
I hadn't thought of the config ontology as an SPI, but I do see indexing as a common repository sidecar, and think it makes sense to encourage implementations to be interoperable.  I agree that indexing is a separate concern from the repository, and another implementation might have no need these entities, or for indexing at all.  But if they do, then they should use the same mechanism for configuring it, and be interoperable with other Fedora components.  For example, if the community supports fcrepo-camel, but UCSD decides to maintain fcrepo-message-consumer instead, they should use the same entities to configure which objects should be indexed, and how.

ajs6f replied:
Right. So we can do that by establishing a standard for indexing
"sidecars". That would include an ontology (much like the one we're now
using) and possibly some common understanding of the workflow implied by
the use of predicates and types in that ontology. Then, when I decide to do
something odd like create an indexing component using Apache UIMA for
feature extraction from images, I can just supply it as a "standardized"
part that will work with UCSD's systems.

On Tue, Jan 20, 2015 at 11:28 AM, Esmé Cowles <notifications@github.com>
wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23234040>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> I hadn't thought of the config ontology as an SPI, but I do see indexing
> as a common repository sidecar, and think it makes sense to encourage
> implementations to be interoperable. I agree that indexing is a separate
> concern from the repository, and another implementation might have no need
> these entities, or for indexing at all. But if they do, then they should
> use the same mechanism for configuring it, and be interoperable with other
> Fedora components. For example, if the community supports fcrepo-camel, but
> UCSD decides to maintain fcrepo-message-consumer instead, they should use
> the same entities to configure which objects should be indexed, and how.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23234040>.
>

acoburn replied:
@ajs6f what would you think of a proposal to publish an indexing# ontology in the fedora.info domain?

Or do you see that as being the responsibility of repository implementations?

ajs6f replied:
I think that's a fine idea. (Although I'm honestly not too clear on the
differences between fcrepo.org, fedora.info, etc.) And if we were to do
that, it would make a lot of sense for the "source" for that published
ontology to be managed here, in its own repo, and paralleled by some
documentation on the wiki explaining what an indexing component should do
to respect the use of that ontology.

On Tue, Jan 20, 2015 at 4:10 PM, Aaron Coburn <notifications@github.com>
wrote:

> In config.rdf
> <https://github.com/fcrepo4/ontology/pull/13#discussion-diff-23256067>:
>
> > +<?xml version="1.0"?>
> > +<?xml-stylesheet type="text/xsl" href="owl2html.xsl"?>
> > +<rdf:RDF
> > +     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
> > +     xmlns:owl="http://www.w3.org/2002/07/owl#"
> > +     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
> > +     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
> > +
> > +    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/config#">
> > +        <rdfs:label xml:lang="en">Fedora Commons Configuration Ontology</rdfs:label>
> > +        <rdfs:comment xml:lang="en">Ontology for user-defined configuration properties and classes on Fedora-curated resources.</rdfs:comment>
> > +    </owl:Ontology>
> > +
> > +    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/config#hasIndexingTransformation">
> > +        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
> > +        <rdfs:domain rdf:resource="http://fcrepo.info/definitions/v4/repository#Container"/>
>
> @ajs6f <https://github.com/ajs6f> what would you think of a proposal to
> publish an indexing# ontology in the fedora.info domain?
>
> Or do you see that as being the responsibility of repository
> implementations?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fcrepo4/ontology/pull/13/files#r23256067>.
>

acoburn replied:
@ajs6f when you write "in its own repo", do you mean separately from the fcrepo4/ontology github repository?

ajs6f replied:
Good point-- I was mixing up the various levels of hierarchy available in Git/GitHub. I suppose I mean that we could either:

- Subdivide this repository with folders for "repository ontologies", "indexing ontologies", etc. This would have to evolve as the ontologies evolve with the HTTP API specs.
- Start a new "indexing SPI" repository and put the indexing ontology there. This is probably better, because there might be additional resources for that SPI, beyond the ontology.

acoburn commented:
As discussed in the [1-22-15 tech call](https://wiki.duraspace.org/display/FF/2015-01-22+-+Fedora+Tech+Meeting), this will be implemented in an indexing# namespace, not a config# namespace.

acoburn commented:
@ajs6f @escowles does this correctly represent what we have discussed? 

Note that this departs from the current use and documentation of the indexing namespace in that I define the owl:class as indexing:Indexable (rather than indexing:indexable).

escowles commented:
:+1:  looks good to me

ajs6f commented:
I'm not seeing anything here that we couldn't do with RDFS. Is there some reason to use OWL instead?

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+
+    <owl:Ontology rdf:about="http://fedora.info/definitions/v4/indexing#">
+        <rdfs:label xml:lang="en">Fedora Commons Indexing Ontology</rdfs:label>
+        <rdfs:comment xml:lang="en">Ontology for user-defined indexing properties and classes on Fedora-curated resources.</rdfs:comment>
```
We might want to use other language than "user-defined", because that's a little ambiguous here. Maybe just something like "Ontology for indexing configuration in Fedora repositories"?.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/indexing#hasIndexingTransformation">
+        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
+        <rdfs:domain rdf:resource="http://fedora.info/definitions/v4/indexing#Indexable"/>
+        <rdfs:range rdf:resource="&xsd;string"/>
```
Do we perhaps want to restrict this to URI?

escowles replied:
I don't think this is a URI, since it's just the name of the transform not the full URI to it.  So we could restrict this to xsd:NMTOKEN if we wanted to be more restrictive than xsd:string.

acoburn replied:
It was my impression that the range was typically a string like so:

    <> indexing:hasIndexingTransformation "default" .

That's how I've used it, and that's how it is described in the [documentation](https://wiki.duraspace.org/display/FEDORA40/External+Search#ExternalSearch-Markaresourceasindexableandassignanappropriateindexingtransformation)

ajs6f replied:
Good point. And other impls might do something very different with this. Maybe we just shouldn't assign a range here at all.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+        <rdfs:comment xml:lang="en">Ontology for user-defined indexing properties and classes on Fedora-curated resources.</rdfs:comment>
+    </owl:Ontology>
+
+    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/indexing#hasIndexingTransformation">
```
Just as a matter of readability, we might want to use a namespace prefix in the "about" here.

acoburn replied:
it was my understanding that in rdf:about fields, you need to use fully qualified URIs (no qnames). It is certainly possible to set an xml:base and then use rdf:ID (instead of rdf:about), but there, too, I thought that rdf:about is preferred over rdf:ID for public ontologies.

ajs6f replied:
No, that's an artifact of the RDF/XML serialization, not inherent in RDF. I don't know that RDF/XML offers us anything here, really, so I'd vote to move to Turtle and use namespace prefixes. I think the whole thing would read better that way.

acoburn replied:
Turtle is certainly more readable (and easier to write). The only issue there is that we have an existing XSL document to transform the ontology into HTML, and we'd be dropping support for that.

ajs6f replied:
Oh, bummer. Well, it might get a little funky, but we could do Turtle -> RDF/XML (just for doc generation) -> HTML (by the same means as now).

escowles replied:
You can define entities in XML, and use those to shorten the URIs.

``` xml
<!DOCTYPE rdf:RDF [
    <!ENTITY indexing "http://fedora.info/definitions/v4/indexing#" >
]>
...
<owl:DatatypeProperty rdf:about="&indexing;hasIndexingTransformation">
```

But I think it's fine to leave the full URIs.  I don't think the shortening makes it much easier to read.  Also, I'd rather stick with RDF/XML, or at least provide RDF/XML as one of a couple formats, since it seems to be the most common format for ontology docs.

awoods replied:
Agreed, I see the value in some degree of consistency in the ontologies hosted in this GitHub repository, including an XML serialization and use of the existing XSL.

ajs6f replied:
Definitely value in consistency, although I don't agree that RDF/XML is a good choice. But it's a pretty minor point. I'm more concerned to see a move to RDFS.

acoburn replied:
Related to this, the existing owl2html.xsl does not work with an RDFS implementation (at least it generates an uninformative HTML file).

ajs6f replied:
Where did we get that? I think we really want to use RDFS here-- we never want to use OWL if we can use something less heavy.

escowles replied:
I wrote it -- and I already have a version for RDFS because I was using that for our updated local ontology: https://github.com/ucsdlib/dams/blob/master/ontology/rdfs2html.xsl

ajs6f replied:
@escowles to the rescue!

acoburn replied:
:clap:

acoburn replied:
Can we include that XSL file in this github repo?

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+
+    <owl:DatatypeProperty rdf:about="http://fedora.info/definitions/v4/indexing#hasIndexingTransformation">
+        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
+        <rdfs:domain rdf:resource="http://fedora.info/definitions/v4/indexing#Indexable"/>
```
See namespace comment above.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+        <rdfs:comment xml:lang="en">Specifies an ld-path indexing transformation</rdfs:comment>
+    </owl:DatatypeProperty>
+
+    <owl:Class rdf:about="http://fedora.info/definitions/v4/indexing#Indexable">
```
See namespace comment above.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+
+    <owl:Class rdf:about="http://fedora.info/definitions/v4/indexing#Indexable">
+        <rdfs:label xml:lang="en">indexable resource</rdfs:label>
+        <rdfs:subClassOf rdf:resource="http://fedora.info/definitions/v4/repository#Resource"/>
```
See namespace comment above.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,27 @@
+        <rdfs:label xml:lang="en">has indexing transformation</rdfs:label>
+        <rdfs:domain rdf:resource="http://fedora.info/definitions/v4/indexing#Indexable"/>
+        <rdfs:range rdf:resource="&xsd;string"/>
+        <rdfs:comment xml:lang="en">Specifies an ld-path indexing transformation</rdfs:comment>
```
I'm not sure we should say anything about LDPath here. That's an fact about the current reference implementation, but maybe we will provide for SPARQL transformations, or another indexing impl might provide for some completely different kind of transform, like XSLT or some kind of process that works over bitstream data.

acoburn replied:
Good point about ldpath. I'll remove the reference to that.

acoburn commented:
@ajs6f RDFS would also work for this -- I used OWL just out of habit.

ajs6f commented:
Let's do RDFS, on the principle of "least power".

acoburn commented:
@ajs6f @escowles This is the latest iteration of the indexing# ontology, implemented as RDFS. Please comment!

acoburn commented on pom.xml:
```diff
@@ -47,6 +45,20 @@
               <transformationSet>
                 <dir>.</dir>
                 <includes>
+                  <include>indexing.rdf</include>
```
This section depends on an `rdfs2html.xsl` file, which was not included in the commit. Should I revert that change or simply include the XSLT from @escowles ?

escowles replied:
@acoburn Go ahead and include that file.

acoburn replied:
@escowles Thanks so much!

awoods replied:
@acoburn, I am not yet seeing the `rdfs2html.xsl` file. Have you yet to commit it or am I missing something?

acoburn replied:
I haven't added it yet. It will come with the next commit -- it needed a bit of modification to work with rdf:Property and rdfs:Class

acoburn replied:
@awoods -- the file is there now.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,33 @@
+    <rdf:Description rdf:about="http://fedora.info/definitions/v4/indexing#">
+      <dcterms:title xml:lang="en">Fedora Commons Indexing Ontology</dcterms:title>
+      <dcterms:modified rdf:datatype="http://www.w3.org/2001/XMLSchema#date">2015-01-23</dcterms:modified>
+      <dcterms:publisher rdf:resource="http://fedora.info"/>
```
Is this a canonical URI for Fedora Commons?

acoburn replied:
@awoods any comment on this?

awoods replied:
I consider `http://fedorarepository.org/` to be the canonical URI for Fedora Commons, but since `http://fedora.info` redirects there and the ontologies are published under `http://fedora.info`, fedora.info may be quite reasonable.
If we do not care about the beauty of aligning the dcterms:publisher with the actual ontology URLs, then I would suggest fedorarepository.org, otherwise it is good as is.

ajs6f replied:
All that is fine with me. For the future, we might want to think about dedicating one of our many domains to "spec stuff", and keeping the others for stuff relating to our reference impl.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,33 @@
+      <rdfs:comment xml:lang="en">Ontology for indexing configuration in Fedora repositories.</rdfs:comment>
+    </rdf:Description>
+
+    <rdf:Description rdf:about="http://fedora.info/definitions/v4/indexing#hasIndexingTransformation">
```
Why did we switch to rdf:Description with type properties? Why not just rdfs:Property as the containing element?

acoburn replied:
I had originally implemented this with rdf:Property, but @escowles XSLT expected rdf:Description (and I usually do everything possible to avoid writing XSLT). That said, I could also just take a deep breath and write some XSL.

ajs6f replied:
Well, if you really don't want to touch it, I can. When everything else in here is addressed (and there isn't much more), I can add a last commit to address this. Or you can. {grin}

acoburn replied:
It's really not that big of a deal. I'll go ahead and include a modified XSL document in the next commit.

ajs6f commented on indexing.rdf:
```diff
@@ -0,0 +1,33 @@
+      <rdfs:isDefinedBy rdf:resource="http://fedora.info/definitions/v4/indexing#"/>
+    </rdf:Description>
+
+    <rdf:Description rdf:about="http://fedora.info/definitions/v4/indexing#Indexable">
```
Why not just rdfs:Class as the containing element?

acoburn replied:
same as the above comment about rdf:Property

acoburn commented:
OK. Let's try this again.

I've added an RDFS implementation of the `indexing#` vocabulary with `rdf:Property` and `rdfs:Class`. I have also added an `rdfs2html.xsl` document, somewhat modified from @escowles version, which included tidying up the HTML so that it matches the output from `owl2html.xsl`. I also split the `owl2html.xsl` into two documents -- keeping all of the owl-specific parsing in the original document, and moving the `/index` template stuff into its own file. Maven, etc has been updated so that

    mvn xml:transform

should work properly.

acoburn commented on rdfs2html.xsl:
```diff
@@ -0,0 +1,145 @@
+      <xsl:value-of select="substring-after(@rdf:about,$about)"/>
+    </xsl:variable>
+    <xsl:if test="$id != ''">
+      <h3 id="{$id}"><xsl:value-of select="$id"/></h3>
```
actually, this should be `<h4>` in order to match the `owl2html` formatting

awoods commented:
Great. Is everyone happy? Are all comments addressed?
If so, I will push it through.
Thanks for working through the details, @acoburn.

ajs6f commented:
I think we're in good shape to move forward. Next week, I'll find something else about which to complain endlessly.

awoods commented:
It would be worrisome if you didn't.

awoods commented:
Resolved with: https://github.com/fcrepo4/ontology/commit/d13033b8402d7c1559f7562dcfa4ab3b7d41dcb0
